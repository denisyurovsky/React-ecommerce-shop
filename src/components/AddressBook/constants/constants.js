export const LISTBOX_PADDING = 8;
export const DOUBLE_LISTBOX_PADDING = LISTBOX_PADDING * 2;
export const LIMIT_NUMBER_ITEMS = 8;
export const ITEM_SIZE_LG = 48;
export const ITEM_SIZE_MD = 36;
