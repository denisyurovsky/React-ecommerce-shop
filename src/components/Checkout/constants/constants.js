export const PANEL = {
  DELIVERY_ADDRESS: 'Delivery Address',
  PERSONAL_INFORMATION: 'Personal Information',
  PERSONAL_ADDRESS: 'Personal Address',
  PAYMENT_METHOD: 'Payment method',
  ORDER_CONFIRMATION: 'Order confirmation',
};
