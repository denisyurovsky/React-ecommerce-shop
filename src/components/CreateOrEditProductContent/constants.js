export const PRICE_ERROR = 'Only numbers';
export const DESCRIPTION_ERROR = 'Only latin characters';
export const LENGTH_ERROR = 'Should contain at least one symbol';
export const DISCOUNT_ERROR = 'Discount should be less than price';
export const SELECTION_ERROR = 'Choose category';
