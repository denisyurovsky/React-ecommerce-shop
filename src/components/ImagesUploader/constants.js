export const LOSS_OF_IMAGES = 'We lost images';

export const STATUS = {
  CROPPED: 'cropped',
  CROPPING: 'cropping',
  IDLE: 'idle',
};

export const TEST_IMG_URL = 'http://someURL';
