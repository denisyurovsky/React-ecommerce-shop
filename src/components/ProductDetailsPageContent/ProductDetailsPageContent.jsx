import PersonIcon from '@mui/icons-material/Person';
import {
  IconButton,
  Typography,
  Box,
  Button,
  Container,
  Card,
  Rating,
} from '@mui/material';
import PropTypes from 'prop-types';
import React from 'react';
import { useSelector } from 'react-redux';

import { formatDate, formatPrice } from '../../helpers/utils/formatData';
import { getRatingByProductId } from '../../store/products/productsSlice';
import { getWishlist } from '../../store/user/userSlice';
import AddToWishListButton from '../AddToWishListButton/AddToWishListButton';
import Carousel from '../Carousel/Carousel';

import Feedback from './Feedback/Feedback';

import styles from './ProductDetailsPageContent.module.scss';

const ProductDetailsPageContent = ({ product }) => {
  const { name, updatedAt, price, images, author, description, id, rating } =
    product;
  const updatedRating = useSelector((state) => getRatingByProductId(state, id));
  const wishlist = useSelector(getWishlist);
  const isWished = (productId, wishlist) => new Set(wishlist).has(productId);

  return (
    <Container>
      <Card className={styles.card}>
        <Box className={styles.title}>
          <Typography variant="h5">{name}</Typography>
          <Typography variant="body1" color="text.secondary">
            {`date: ${formatDate(updatedAt)}`}
          </Typography>
        </Box>
        <Typography variant="h4" className={styles.price}>
          {formatPrice(price)}
        </Typography>
        <Box className={styles.images}>
          <Carousel images={images} />
          <AddToWishListButton
            productId={id}
            productName={name}
            isAddedToWishlist={isWished(id, wishlist)}
          />
        </Box>
        <Box className={styles.buy}>
          <Rating
            data-testid="total-rating"
            value={updatedRating ?? rating}
            sx={{ mt: 4 }}
            size="large"
            readOnly
            precision={0.5}
          />
          <Button variant="contained" sx={{ mt: 3, mb: 4 }}>
            + ADD TO CART
          </Button>
          <Box className={styles.seller}>
            <Typography variant="body1">
              {`${author.firstName} ${author.lastName}`}
            </Typography>
            <IconButton className={styles.personIcon}>
              <PersonIcon />
            </IconButton>
          </Box>
        </Box>
        <Box className={styles.description}>
          <Typography variant="body1" color="text.secondary">
            {description}
          </Typography>
        </Box>
      </Card>
      <Feedback productId={id} />
    </Container>
  );
};

ProductDetailsPageContent.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.number.isRequired,
    rating: PropTypes.number,
    name: PropTypes.string,
    updatedAt: PropTypes.string,
    price: PropTypes.number,
    images: PropTypes.arrayOf(PropTypes.string),
    author: PropTypes.shape({
      firstName: PropTypes.string,
      lastName: PropTypes.string,
    }),
    description: PropTypes.string,
  }).isRequired,
};

export default ProductDetailsPageContent;
