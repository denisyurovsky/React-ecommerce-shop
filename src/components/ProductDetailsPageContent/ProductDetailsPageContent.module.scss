@import "../../assets/styles/variables.scss";
@import "../../assets/styles/colors.scss";

$productDetailsPageContentHeight: calc(100% - 40px);

.container {
  margin: 20px auto;
  display: grid;
  grid-template-columns: auto 220px;
  grid-template-rows: 80px auto auto;
  align-items: center;
  border-radius: 2px;
  box-shadow: 0 0 2px $greyDark;
  height: $productDetailsPageContentHeight;
  width: 100%;
  background: $white;

  @media screen and (max-width: $sm) {
    grid-template-columns: 2fr 1fr;
    grid-template-rows: 90px auto 170px auto;
    height: fit-content;
  }
}

.container > * {
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.title {
  grid-area: 1 / 1 / 2 / 2;
  align-items: flex-start;
}

.title > * {
  padding-left: 1.6rem;
}

.price {
  grid-area: 1 / 2 / 2 / 3;
}

.images {
  grid-area: 2 / 1 / 3 / 2;
  border-top: 1px solid $greyDark;
  border-bottom: 1px solid $greyDark;
  border-right: 1px solid $greyDark;

  @media screen and (max-width: $sm) {
    grid-area: 2 / 1 / 3 / 3;
    border-right: none;
    border-bottom: none;
  }
}

.buy {
  grid-area: 2 / 2 / 3 / 3;
  border-top: 1px solid $greyDark;
  border-bottom: 1px solid $greyDark;
  justify-content: space-evenly;

  .seller {
    display: flex;
    flex-direction: row;
    align-items: center;

    .name {
      margin: 0 0.5rem 0 1rem;
      text-align: center;
      width: fit-content;
    }

    .personIcon {
      color: $white;
      background: $greyDark !important;
      margin: 0 1rem 0 0.5rem;
    }
  }

  @media screen and (max-width: $sm) {
    grid-area: 3 / 1 / 4 / 3;
  }
}

.description {
  grid-area: 3 / 1 / 4 / 3;
  padding: 35px;

  @media screen and (max-width: $sm) {
    grid-area: 4 / 1 / 5 / 3;
  }
}
