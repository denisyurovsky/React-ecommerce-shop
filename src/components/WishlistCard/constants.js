export const DEFAULT_LABEL_INPUT = 'Wishlist Name';
export const ERROR_LABEL_INPUT = 'Wishlist name should be unique';
