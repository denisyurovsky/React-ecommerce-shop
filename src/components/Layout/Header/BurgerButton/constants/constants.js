import { pathNames } from '../../../../../constants/pathNames';

export const LINKS = [
  {
    name: 'HOME',
    link: '/',
  },
  {
    name: 'PRODUCTS',
    link: pathNames.PRODUCTS,
  },
];
