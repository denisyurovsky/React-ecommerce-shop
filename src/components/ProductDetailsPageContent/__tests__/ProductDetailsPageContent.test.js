import { rest } from 'msw';
import { setupServer } from 'msw/node';
import React from 'react';

import { DEFAULT_NAME } from '../../../constants/feedbackConstants';
import convertDescription from '../../../helpers/convertDescriptionToObj';
import { testCart } from '../../../test-utils/dto/cartDto';
import feedbackDto from '../../../test-utils/dto/feedbackDto';
import productsDto from '../../../test-utils/dto/productsDto';
import usersDto from '../../../test-utils/dto/usersDto';
import { handleModal } from '../../../test-utils/feedback/feedbackHandlers';
import render, { screen } from '../../../test-utils/renderWith';
import ProductDetailsPageContent from '../ProductDetailsPageContent';

const handlersFulfilled = [
  rest.get('/products/1', (req, res, ctx) => res(ctx.json(productsDto[0]))),
  rest.get('/feedbacks', (req, res, ctx) => res(ctx.json([]))),
];

const addFeedbackHandlers = [
  rest.get('/feedbacks', (req, res, ctx) => {
    const productId = req.url.searchParams.get('productId');

    switch (productId.toString()) {
      case '2':
        return res(ctx.json(feedbackDto.slice(0, 2)));
      case '1':
        return res(ctx.json([feedbackDto[2]]));
    }
  }),
  rest.post('/feedbacks', (req, res, ctx) =>
    res(
      ctx.json({
        ...req.body,
        id: 3,
        createdAt: '2022-01-02T08:54:18.560Z',
      })
    )
  ),
  rest.patch('/rating', (req, res, ctx) =>
    res(
      ctx.json({
        ...req.body,
        id: req.body.productId,
      })
    )
  ),
];

const initialUser = {
  user: { id: 1 },
};

describe('ProductDetailsPageContent', () => {
  describe('snapshots', () => {
    const server = setupServer(...handlersFulfilled);

    beforeAll(() => server.listen());
    afterAll(() => server.close());

    it('renders a valid snapshot', async () => {
      const product = {
        ...productsDto[0],
        description: convertDescription(productsDto[0].description),
      };
      const { asFragment, findByTestId } = render(
        <ProductDetailsPageContent product={product} />,
        { cart: testCart, user: initialUser }
      );

      await findByTestId('comments');
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Product rating', () => {
  const server = setupServer(...addFeedbackHandlers);

  beforeAll(() => server.listen());
  afterAll(() => server.close());

  it('should show total rating', async () => {
    const product = {
      ...productsDto[2],
      description: convertDescription(productsDto[2].description),
    };
    const { getByTestId } = render(
      <ProductDetailsPageContent product={product} />,
      { cart: testCart, user: initialUser }
    );

    await screen.findByTestId('comments');

    expect(
      getByTestId('total-rating').querySelectorAll('[data-testid="StarIcon"]')
    ).toHaveLength(8);
  });

  it('should change total rating', async () => {
    const product = {
      ...productsDto[2],
      description: convertDescription(productsDto[2].description),
    };

    render(<ProductDetailsPageContent product={product} />, {
      cart: testCart,
      user: usersDto[3],
    });

    await screen.findByTestId('comments');
    handleModal({ rating: 2 });

    await screen.findByText(DEFAULT_NAME, { exact: false });
    await screen.findAllByTestId('StarBorderIcon');

    expect(
      screen
        .getByTestId('total-rating')
        .querySelectorAll('[data-testid="StarIcon"]')
    ).toHaveLength(7);
  });
});
