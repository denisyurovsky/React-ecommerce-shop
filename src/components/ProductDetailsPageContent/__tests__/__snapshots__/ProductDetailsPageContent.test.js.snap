// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetailsPageContent snapshots renders a valid snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="title MuiBox-root css-0"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Intelligent Cotton Pants
      </h5>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
      >
        date: 24.11.2021
      </p>
    </div>
    <h4
      class="MuiTypography-root MuiTypography-h4 price css-5lbw0b-MuiTypography-root"
    >
      670Â $
    </h4>
    <div
      class="images MuiBox-root css-0"
    >
      <div
        class="container single"
        data-testid="left"
      >
        <div
          aria-live="polite"
          class="image-gallery"
        >
          <div
            class="image-gallery-content left"
          >
            <div
              class="image-gallery-slide-wrapper left"
            >
              <div
                class="image-gallery-slides"
              >
                <div
                  aria-label="Go to Slide 1"
                  class="image-gallery-slide  center  single"
                  role="button"
                  style="display: inherit; transform: translate3d(0%, 0, 0);"
                  tabindex="-1"
                >
                  <img
                    alt=""
                    class="image-gallery-image"
                    height="400px"
                    sizes=""
                    src="noImg.png"
                    srcset=""
                    title=""
                    width=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="buy MuiBox-root css-0"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        + ADD TO CART
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="seller MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 name css-ahj2mt-MuiTypography-root"
        >
          Lindsay Yundt
        </p>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium personIcon css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="description MuiBox-root css-0"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
      >
        The Apollotech B340 is an affordable wireless mouse with reliable connectivity, 12 months battery life and modern design
      </p>
    </div>
  </div>
</DocumentFragment>
`;
